[mlir][Linalg] Refine how broadcast dims are treated

banach-space opened the pull request:
This PR fixes how broadcast dims (identified as "zero" results in
permutation maps) corresponding to a reduction iterator are vectorised
in the case of generic Ops. Here's an example:

```mlir
  #map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
  #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, 0)>

  func.func @generic_with_reduction_and_broadcast(%arg0: tensor<1x12x197x197xf32>) -> (tensor<1x12x197x1xf32>) {
    %0 = tensor.empty() : tensor<1x12x197x1xf32>

    %1 = linalg.generic {indexing_maps = [#map, #map1],
                        iterator_types = ["parallel", "parallel", "parallel", "reduction"]}
      ins(%arg0 : tensor<1x12x197x197xf32>)
      outs(%0 : tensor<1x12x197x1xf32>) {

    ^bb0(%in: f32, %out: f32):
      %818 = arith.addf %in, %out : f32
      linalg.yield %818 : f32
    } -> tensor<1x12x197x1xf32>
    return %1 : tensor<1x12x197x1xf32>
  }
```

This is a perfectly valid Generic Op, but currently triggers two issues
in the vectoriser. The root cause is this map:

```mlir
  #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, 0)>
```

This map triggers an assert in `reindexIndexingMap` -  this hook
incorrectly assumes that every result in the input map is a `dim`
expression and that there are no constants. That's not the case in this
example. `reindexIndexingMap` is extended to allow maps like the one
above. For now, only constant "zero" results are allowed. This can be
extended in the future once a good motivating example is available.

Separately, the permutation map highlighted above "breaks" mask
calculation (ATM masks are always computed, even in the presence of
static shapes). When applying the following permutation:
```mlir
  (d0, d1, d2, d3) -> (d0, d1, d2, 0)
```

to these canonical shapes (corresponding to the example above):
```
  (1, 12, 197, 197)
```
we end up with the following error:
```bash
error: vector types must have positive constant sizes but got 1, 12, 197, 0
```

The error makes sense and indicates that we should update the
permutation map above to:
```
  (d0, d1, d2, d3) -> (d0, d1, d2)
```

This would correctly give the following vector type:
```
  vector<1x12x197xi1>
```

Fixes #97247

banach-space committed:
[mlir][Linalg] Refine how broadcast dims are treated

This PR fixes how broadcast dims (identified as "zero" results in
permutation maps) corresponding to a reduction iterator are vectorised
in the case of generic Ops. Here's an example:

```mlir
  #map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
  #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, 0)>

  func.func @generic_with_reduction_and_broadcast(%arg0: tensor<1x12x197x197xf32>) -> (tensor<1x12x197x1xf32>) {
    %0 = tensor.empty() : tensor<1x12x197x1xf32>

    %1 = linalg.generic {indexing_maps = [#map, #map1],
                        iterator_types = ["parallel", "parallel", "parallel", "reduction"]}
      ins(%arg0 : tensor<1x12x197x197xf32>)
      outs(%0 : tensor<1x12x197x1xf32>) {

    ^bb0(%in: f32, %out: f32):
      %818 = arith.addf %in, %out : f32
      linalg.yield %818 : f32
    } -> tensor<1x12x197x1xf32>
    return %1 : tensor<1x12x197x1xf32>
  }
```

This is a perfectly valid Generic Op, but currently triggers two issues
in the vectoriser. The root cause is this map:

```mlir
  #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, 0)>
```

This map triggers an assert in `reindexIndexingMap` -  this hook
incorrectly assumes that every result in the input map is a `dim`
expression and that there are no constants. That's not the case in this
example. `reindexIndexingMap` is extended to allow maps like the one
above. For now, only constant "zero" results are allowed. This can be
extended in the future once a good motivating example is available.

Separately, the permutation map highlighted above "breaks" mask
calculation (ATM masks are always computed, even in the presence of
static shapes). When applying the following permutation:
```mlir
  (d0, d1, d2, d3) -> (d0, d1, d2, 0)
```

to these canonical shapes (corresponding to the example above):
```
  (1, 12, 197, 197)
```
we end up with the following error:
```bash
error: vector types must have positive constant sizes but got 1, 12, 197, 0
```

The error makes sense and indicates that we should update the
permutation map above to:
```
  (d0, d1, d2, d3) -> (d0, d1, d2)
```

This would correctly give the following vector type:
```
  vector<1x12x197xi1>
```

Fixes #97247

MaheshRavishankar commented:
I have not looked into details of this PR, and there is value in having the lower levels of the stack being robust, but 
```
 #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, 0)>
```
is not the canonical representation of broadcast IMO. This is.

```
 #map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
```

So just signalling that here.

MacDue commented on mlir/include/mlir/IR/AffineMap.h:
```diff
@@ -346,6 +346,10 @@ class AffineMap {
   /// returns the resulting values. `this` must be symbol-less.
   SmallVector<int64_t, 4> compose(ArrayRef<int64_t> values) const;
 
+  size_t numOfZeroResults() const;
+
+  AffineMap dropZeros();
```
Missing doc comments

dcaballe replied:
Could you please check if we have a utility for this already? As I mentioned in another PR, we have so many AffineMap utilities that sometimes it's not easy to find what we need :)

banach-space replied:
I saw your comment on the other PR and that immediately made me think about this :)

I couldn't find anything that would help for these specific case. However, we do have:
* [getBroadcastDims()](https://github.com/llvm/llvm-project/blob/867ff2d4268ca7eed89a24f100b67b68f5329439/mlir/include/mlir/IR/AffineMap.h#L149-L155)
  * NOTE: to be removed in https://github.com/llvm/llvm-project/pull/99341/,
* [isMinorIdentityWithBroadcasting()](https://github.com/llvm/llvm-project/blob/867ff2d4268ca7eed89a24f100b67b68f5329439/mlir/include/mlir/IR/AffineMap.h#L157-L164),
* [dropResults()](https://github.com/llvm/llvm-project/blob/867ff2d4268ca7eed89a24f100b67b68f5329439/mlir/include/mlir/IR/AffineMap.h#L297-L307).

So, there's scope for re-use. 

Here's what I suggest:
*  Introduce `SmallVector<unsigned> getZeroResults()`
   * It would return the positions of **zero**/**bcast** results.
* Simplify `isMinorIdentityWithBroadcasting()`
  * It would use the newly introduced `getZeroResults()`.
* Replace `dropZeros()` (proposed in this PR) with a call to `dropResults(getZeroResults())`.

I would still have to introduce `getZeroResults()` (i.e. yet another hook in this file), but would avoid introducing `dropZeros()` :) WDYT?

dcaballe replied:
If there is nothing that provides what you need I would just introduce what you need and try to reuse/refactor at implementation level.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -629,7 +629,21 @@ static Value buildVectorWrite(RewriterBase &rewriter, Value value,
+  //    vector<1x16x16xty>
+  // rather than:
+  //    vector<1x16x16x0xty>
+  auto opOperantMapWithoutZeros = opOperandMap.dropZeros();
```
Nit: expand auto plz.

banach-space replied:
Done

ftynse commented on mlir/lib/IR/AffineMap.cpp:
```diff
@@ -540,6 +540,18 @@ AffineMap AffineMap::dropResults(const llvm::SmallBitVector &positions) const {
 }
 
+AffineMap AffineMap::dropZeros() {
+  auto exprs = llvm::to_vector<4>(getResults());
+  SmallVector<AffineExpr, 8> newExprs;
```
```suggestion
  auto exprs = llvm::to_vector(getResults());
  SmallVector<AffineExpr> newExprs;
```

ftynse commented on mlir/lib/IR/AffineMap.cpp:
```diff
@@ -540,6 +540,18 @@ AffineMap AffineMap::dropResults(const llvm::SmallBitVector &positions) const {
+  SmallVector<AffineExpr, 8> newExprs;
+
+  for (auto expr : getResults()) {
+    auto constExpr = dyn_cast<AffineConstantExpr>(expr);
+    if (!constExpr)
```
This will ignore _any_ constant, not just zero as the name implies.

banach-space replied:
Fixed, thanks!

dcaballe commented on mlir/lib/IR/AffineMap.cpp:
```diff
@@ -579,6 +591,17 @@ SmallVector<int64_t, 4> AffineMap::compose(ArrayRef<int64_t> values) const {
   return res;
 }
 
+size_t AffineMap::numOfZeroResults() const {
```
nit: `getNum...`

dcaballe commented on mlir/lib/IR/AffineMap.cpp:
```diff
@@ -540,6 +540,18 @@ AffineMap AffineMap::dropResults(const llvm::SmallBitVector &positions) const {
   return AffineMap::get(getNumDims(), getNumSymbols(), exprs, getContext());
 }
 
+AffineMap AffineMap::dropZeros() {
```
nit: `dropZeroResults`?

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -629,7 +629,21 @@ static Value buildVectorWrite(RewriterBase &rewriter, Value value,
+  //    vector<1x16x16x0xty>
+  auto opOperantMapWithoutZeros = opOperandMap.dropZeros();
+  write =
+      state.maskOperation(rewriter, write, linalgOp, opOperantMapWithoutZeros);
```
What happens with the xfer read counterpart? Should we move this logic into `maskOperation`?

banach-space replied:
> What happens with the xfer read counterpart? 

It turns out you've already implemented that :) 

https://github.com/llvm/llvm-project/blob/2ba3fe7356f065757a2279f65e4ef5c8f1476293/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1327-L1335

> Should we move this logic into maskOperation?

Yes! Sending an update shortly.

dcaballe reviewed (commented):
Thanks for fixing this! Left a few comments.

I would also revisit the commit description as this seems to be a reduction problem, not a broadcast problem? Or maybe both broadcast and reduction?

banach-space committed:
fixup! [mlir][Linalg] Refine how broadcast dims are treated

Addressing PR comments

banach-space committed:
fixup! [mlir][Linalg] Refine how broadcast dims are treated

* Move the logic to remove zero from indexing maps to `maskOperation`
* Update the input mask name in `maskOperation` to `maybeIndexingMap` -
  the actual input is always an indexing map extracted from the
  corresponding linalg Op
* Remove the duplicated comment for `maskOperation`

banach-space commented:
Ping :)

banach-space commented:
@dcaballe , could you take another look?

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -224,10 +224,10 @@ struct VectorizationState {
   /// is not needed. If provided, the canonical mask for this operation is
-  /// permuted using `maybeMaskingMap`.
+  /// permuted using `maybeIndexingMap`.
   Operation *
   maskOperation(RewriterBase &rewriter, Operation *opToMask, LinalgOp linalgOp,
-                std::optional<AffineMap> maybeMaskingMap = std::nullopt);
+                std::optional<AffineMap> maybeIndexingMap = std::nullopt);
```
I think indexing map would be misleading here as it's a term very tight to linalg ops and here we may apply it to arbitrary ops. Both will evolve in different ways for more complex vectorization scenarios, like those with control flow.

banach-space replied:
Thanks for reminding me about this!

Right now we are actually passing the indexing map whenever this hook is used - this is just to documenting the state today. So let me leave it as is.

I will send a follow-on PR in which I'll try to clarify the difference between the two maps.

dcaballe commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1317,16 +1330,6 @@ vectorizeAsLinalgGeneric(RewriterBase &rewriter, VectorizationState &state,
-      }
-    }
-    AffineMap maskingMap = indexingMap.dropResults(zeroPos);
-
```
Oh, ok, we are just moving this into `maskOperation` because we have other entry points where this was not executed... Got it.
